### PR TITLE
New inner_combine_linter

### PR DESCRIPTION
**Pull request by MichaelChirico**
Part of #962 

This one is somewhat of a mess... easy enough to describe what it wants to do, and it did match a bunch of our internal code. But a bit of a mess to write the linter.

### CODE REVIEW

File: R/inner_combine_linter.R [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    #   but writing the xpath for the alternative case was proving too tricky.
+    #   It's messy enough as is -- it may make sense to take another pass at
+    #   writing the xpath from scratch to see if it can't be simplified.
+
+    # See ?as.Date, ?as.POSIXct. tryFormats is not explicitly in any default
+    #   POSIXct method, but it is in as.Date.character and as.POSIXlt.character --
+    #   the latter is what actually gets invoked when running as.POSIXct
+    #   on a character. So it is indeed an argument by pass-through.
+    date_args <- c("format", "origin", "tz", "tryFormats")
+    date_funs <- c("as.Date", "as.POSIXct", "as.POSIXlt")
```
Comment (AshesITR) [inline R/inner_combine_linter.R:38]: We make heavy use of `lubridate`, could add `ymd` and friends here.

File: R/inner_combine_linter.R [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        xp_text_in_table(c(no_arg_vectorized_funs, date_funs, log_funs))
+      ),
+      "not(following-sibling::expr[not(expr[SYMBOL_FUNCTION_CALL])])",
+      sprintf(
+        "not(%1$s != following-sibling::expr/%1$s)",
+        "expr/SYMBOL_FUNCTION_CALL"
+      ),
+      date_args_cond,
+      log_args_cond
+    )
```
Comment (AshesITR) [inline R/inner_combine_linter.R:67]: `glue()` these as well?
Very hard to read with the nested `sprintf()` calls.

File: R/inner_combine_linter.R [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        )
+        paste("Combine inputs to vectorized functions first to take full advantage of vectorization, e.g.,", message)
+      },
+      type = "warning"
+    ))
+  })
+}
+
+' Make the XPath condition ensuring an argument matches across calls
+#'
```
Comment (AshesITR) [inline R/inner_combine_linter.R:96]: Missing `#`.

File: R/inner_combine_linter.R (hunk 1 above)
Comment (MichaelChirico) [reply R/inner_combine_linter.R:38]: Would you mind enumerating which functions? I am not really familiar with `lubridate()`. Or we could add TODO.

Comment (AshesITR) [reply R/inner_combine_linter.R:38]: Sure.
They can be found all together in the two help pages `?ymd` and `?ymd_hms`.

```
ymd(
  ...,
  quiet = FALSE,
  tz = NULL,
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

ydm(
  ...,
  quiet = FALSE,
  tz = NULL,
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

mdy(
  ...,
  quiet = FALSE,
  tz = NULL,
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

myd(
  ...,
  quiet = FALSE,
  tz = NULL,
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

dmy(
  ...,
  quiet = FALSE,
  tz = NULL,
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

dym(
  ...,
  quiet = FALSE,
  tz = NULL,
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

yq(..., quiet = FALSE, tz = NULL, locale = Sys.getlocale("LC_TIME"))

ym(..., quiet = FALSE, tz = NULL, locale = Sys.getlocale("LC_TIME"))

my(..., quiet = FALSE, tz = NULL, locale = Sys.getlocale("LC_TIME"))

ymd_hms(
  ...,
  quiet = FALSE,
  tz = "UTC",
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

ymd_hm(
  ...,
  quiet = FALSE,
  tz = "UTC",
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

ymd_h(
  ...,
  quiet = FALSE,
  tz = "UTC",
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

dmy_hms(
  ...,
  quiet = FALSE,
  tz = "UTC",
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

dmy_hm(
  ...,
  quiet = FALSE,
  tz = "UTC",
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

dmy_h(
  ...,
  quiet = FALSE,
  tz = "UTC",
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

mdy_hms(
  ...,
  quiet = FALSE,
  tz = "UTC",
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

mdy_hm(
  ...,
  quiet = FALSE,
  tz = "UTC",
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

mdy_h(
  ...,
  quiet = FALSE,
  tz = "UTC",
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

ydm_hms(
  ...,
  quiet = FALSE,
  tz = "UTC",
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

ydm_hm(
  ...,
  quiet = FALSE,
  tz = "UTC",
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)

ydm_h(
  ...,
  quiet = FALSE,
  tz = "UTC",
  locale = Sys.getlocale("LC_TIME"),
  truncated = 0
)
```

Comment (MichaelChirico) [reply R/inner_combine_linter.R:38]: Nice, thanks!

File: R/inner_combine_linter.R (hunk 2 above)
Comment (MichaelChirico) [reply R/inner_combine_linter.R]: Refactored a little bit, hopefully a bit better.

File: tests/testthat/test-inner_combine_linter.R [hunk 4]
```diff
@@ -0,0 +1,95 @@
+test_that("inner_combine_linter lints a false positive-ish usage", {
+  # By default as.POSIXct.character picks up the format to apply from
+  #   the first element and, since it succeeds, applies that to the remaining
+  #   timestamps. Whereas when run individually, it won't succeed until
+  #   the correct format is matched for each input. Nevertheless, it is
+  #   still preferable to vectorize the call, while being sure to use a
+  #   consistent format for the inputs. In this case, the correct equivalent
+  #   call is as.POSIXct(c("2021-01-01 00:00:00", "2021-01-01 01:00:00")).
+  # See http://google3/analysis/economics/r/izeitgeist/tests/testthat/test-izeitgeist.R;l=1058-1062;rcl=394984377.
```
Comment (AshesITR) [inline tests/testthat/test-inner_combine_linter.R:9]: internal google link?

**Review by AshesITR — APPROVED** [review]

**Pull request merged**